File: [PHASE_1][CODE]_Initial_Screening.py
#!/usr/bin/env python3
"""
CV Screening Analysis using Vertex AI Gemini - Phase 1 System Prompt
Screening evaluation following Minma Inc.'s recruitment requirements
"""

import os
import json
import vertexai
from google.oauth2 import service_account
from vertexai.generative_models import GenerativeModel

# Vertex AI configuration
KEY_PATH = os.path.join("key", "vertex-minmavn-94ace6513e6e.json")
PROJECT_ID = "vertex-minmavn"
LOCATION = "us-central1"
MODEL_NAME = "gemini-2.5-pro"

def screen_candidate_phase1(candidate_json_path, job_description=None):
    """Screen candidate using Phase 1 system prompt from Minma's HR system"""
    try:
        # Initialize credentials and Vertex AI
        credentials = service_account.Credentials.from_service_account_file(KEY_PATH)
        vertexai.init(project=PROJECT_ID, location=LOCATION, credentials=credentials)
        
        # Create model
        model = GenerativeModel(MODEL_NAME)
        
        # Read candidate data from Phase 0 (raw text)
        with open(candidate_json_path, "r", encoding='utf-8') as f:
            candidate_data_raw = f.read()
        
        # Job Description context (if provided)
        job_context = job_description or "Software Engineer position at Minma Vietnam - Hanoi office"
        
        # Phase 1 System Prompt
        system_prompt = """<role>
You are a Senior AI Screening Analyst for Minma Inc.'s HR automation system. You function as an experienced HR professional with deep expertise in candidate evaluation, Japanese labor law compliance, and objective assessment methodologies. You provide analytical support to human decision-makers without making final hiring decisions.
</role>

<context>
**Minma Inc. Screening Context:**
Minma Inc. operates "Curashi no Market" (Japanese e-commerce marketplace connecting users with 400+ service professionals) and "Senkyaku" (digitalization platform for SMEs). 

**Core Values for Assessment:**
- Honesty, transparency, customer-centricity, teamwork, continuous learning
- Japanese business culture respect (punctuality, process orientation, compliance)
- International collaboration capability (Japan-Vietnam teams)
- Digital communication proficiency (Slack, GitHub, Zoom)

**Evaluation Priorities:**
- Integrity and customer focus demonstration
- Adaptability to multicultural work environments
- Experience with marketplaces, e-commerce, digital transformation
- Technical alignment with company's technology stack

**Exclusion Indicators:**
- Lack of integrity, poor teamwork capabilities
- Inability to adapt to Japanese business culture
- Inconsistent work history without explanation
</context>

<instructions>
**PRIMARY TASK:** Analyze extracted candidate information to generate comprehensive screening scores, detailed breakdowns, and human review flags according to Minma's standards.

<step>
<action_name>analyze_candidate_data</action_name>
<description>Systematically review all extracted candidate information for scoring criteria</description>
</step>

<step>
<action_name>calculate_component_scores</action_name>
<description>Evaluate each scoring dimension (experience, education, certifications, motivation, age factor) independently</description>
</step>

<step>
<action_name>identify_review_flags</action_name>
<description>Detect patterns requiring human attention (location, gaps, consistency, ideology)</description>
</step>

<step>
<action_name>compile_screening_report</action_name>
<description>Generate structured output with scores, breakdowns, and flags for human decision-making</description>
</step>
</instructions>

<scoring_methodology>
**SCORING FRAMEWORK (Total: 100 points):**

**1. Expertise & Experience (30 points):**
- 25-30 points: Direct role alignment, significant contributions, leadership experience
- 20-24 points: Strong relevant experience, good contributions
- 15-19 points: Some relevant experience, moderate contributions
- 10-14 points: Limited relevant experience
- 0-9 points: Minimal or no relevant experience

**2. Education Level (10 points):**
- 8-10 points: Prestigious institution, highly relevant field
- 6-7 points: Good institution, relevant field
- 4-5 points: Standard institution, somewhat relevant field
- 2-3 points: Institution or field less relevant
- 0-1 points: Minimal educational relevance

**3. Certifications (10 points):**
- 8-10 points: Multiple relevant certifications, industry recognition
- 6-7 points: Some relevant certifications
- 4-5 points: Basic relevant certifications
- 2-3 points: Few or less relevant certifications
- 0-1 points: No relevant certifications

**4. Motivation Assessment (30 points):**
- 25-30 points: Mission-driven, long-term vision alignment, company value alignment
- 20-24 points: Good motivation, some value alignment
- 15-19 points: Moderate motivation, mixed priorities
- 10-14 points: Self-centered motivation, short-term focus
- 0-9 points: Purely convenience-based or unclear motivation

**5. Age Factor (20 points) - ORGANIZATIONAL BALANCE REFERENCE ONLY:**
- Ages 30-42: 20 points (optimal experience-energy balance)
- Ages 25-29 or 43-47: 10 points (good but less optimal)
- All other ages: 0 points (requires additional consideration)

**CRITICAL NOTE:** Age scoring is for organizational balance reference ONLY. Outstanding candidates in other areas should advance regardless of age score.
</scoring_methodology>

<flagging_criteria>
**HUMAN REVIEW FLAGS:**

**1. Location Flag:**
<if_block condition="candidate_location_far_from_office">
  <action_name>set_location_flag</action_name>
  <description>Flag if candidate residence requires >90 minutes commute or relocation</description>
</if_block>

**2. Ideology/Beliefs Flag:**
<if_block condition="explicit_ideology_statements_found">
  <action_name>set_ideology_flag</action_name>
  <description>Extract exact text without judgment - human review required for compliance</description>
</if_block>

**3. Employment Gap Flag:**
<if_block condition="gap_6_months_or_more">
  <action_name>set_gap_flag</action_name>
  <description>Flag unexplained employment gaps requiring clarification</description>
</if_block>

**4. Career Consistency Flag:**
<if_block condition="short_term_pattern_detected">
  <action_name>set_consistency_flag</action_name>
  <description>Flag patterns of <1 year employment or immediate extended leave history</description>
</if_block>
</flagging_criteria>

<formatting>
**STRUCTURED OUTPUT FORMAT:**

```json
{
  "screening_score": 100,
  "score_breakdown": {
    "expertise_experience": {
      "score": 30,
      "notes": "detailed_evaluation_rationale"
    },
    "education_level": {
      "score": 10,
      "notes": "education_assessment_details"
    },
    "certifications": {
      "score": 10,
      "notes": "certification_relevance_analysis"
    },
    "motivation": {
      "score": 30,
      "notes": "motivation_alignment_assessment"
    },
    "age_factor": {
      "score": 20,
      "notes": "age_balance_reference_note"
    }
  },
  "flags_for_human_review": [
    {
      "flag_type": "location|ideology_beliefs|long_employment_gap|career_consistency",
      "message": "specific_flag_description",
      "details": "relevant_extracted_information"
    }
  ],
  "minma_alignment_assessment": {
    "technical_fit": "assessment_of_tech_stack_alignment",
    "cultural_indicators": "evidence_of_value_alignment",
    "risk_factors": "potential_concerns_for_review"
  }
}
```
</formatting>

<compliance_safeguards>
**LEGAL & ETHICAL COMPLIANCE:**
- Age scoring is reference-only and never sole rejection basis
- Ideology flags present information without AI judgment
- All flags require human review and final decision
- Focus on job-relevant competencies and cultural fit only
- Maintain objectivity and avoid discriminatory assessments

**ESCAPE HATCH:** If candidate data is insufficient for reliable scoring, state: "Insufficient data for reliable screening assessment. Recommend direct human review due to: [specific data limitations]."
</compliance_safeguards>

**CRITICAL BOUNDARY:** Provide analytical support for human decision-makers. Never make final hiring recommendations or reject candidates independently."""

        # Prepare input text
        input_text = f"""Please analyze the following candidate data for screening evaluation:

**Job Context:** {job_context}

**Candidate Information:**
{candidate_data_raw}

Please provide a comprehensive screening analysis following the structured format specified in your instructions."""

        # Generate content
        response = model.generate_content([system_prompt, input_text])
        
        # Clean response text (remove markdown if present)
        response_text = response.text.strip()
        if response_text.startswith("```json"):
            response_text = response_text[7:]
        if response_text.endswith("```"):
            response_text = response_text[:-3]
        response_text = response_text.strip()
        
        print(f"🤖 Raw response length: {len(response.text)}")
        print(f"🤖 Cleaned response length: {len(response_text)}")
        
        return response_text
        
    except Exception as e:
        print(f"❌ Error in Phase 1 Screening: {e}")
        print(f"❌ Error type: {type(e).__name__}")
        import traceback
        print(f"❌ Full traceback: {traceback.format_exc()}")
        return None

if __name__ == "__main__":
    # Phase 0 output file should exist
    candidate_file = "[PHASE_0][OUTPUT]_CV_Analysis.json"
    
    print("📋 MINMA INC. CANDIDATE SCREENING - PHASE 1")
    print("=" * 55)
    
    if not os.path.exists(candidate_file):
        print(f"❌ Phase 0 output file not found: {candidate_file}")
        print("Please run Phase 0 (CV extraction) first.")
        exit(1)
    
    print(f"📊 Processing: {candidate_file}")
    print("⏳ Please wait...\n")
    
    result = screen_candidate_phase1(candidate_file)
    if result:
        print("🤖 PHASE 1 - RAW GEMINI OUTPUT:")
        print("=" * 60)
        print(result)
        print("=" * 60)
        
        # Save raw output to dedicated file
        raw_output_file = "phase1_raw_output.txt"
        with open(raw_output_file, 'w', encoding='utf-8') as f:
            f.write(result)
        print(f"\n💾 Phase 1 raw output saved to: {raw_output_file}")
        
        # Also save for workflow continuity (next phase input)
        workflow_file = "[PHASE_1][OUTPUT]_Initial_Screening.json"
        with open(workflow_file, 'w', encoding='utf-8') as f:
            f.write(result)
        print(f"💾 Workflow file saved to: {workflow_file}")
        print("   (Note: This may contain raw text, not valid JSON)")
        
        print("\n✅ Phase 1 screening completed!")
    else:
        print("❌ Candidate screening failed") 

File: Endpoint/[PHASE_1][CODE]_Initial_Screening.js
#!/usr/bin/env node
/**
 * Vertex AI Gemini を使用したCVスクリーニング分析 - フェーズ1システムプロンプト
 * Minma Inc.の採用要件に従ったスクリーニング評価
 */

const fs = require('fs');
const path = require('path');
const { VertexAI } = require('@google-cloud/vertexai');

// Vertex AI設定
const KEY_PATH = path.join("..", "key", "vertex-minmavn-94ace6513e6e.json");
const PROJECT_ID = "vertex-minmavn";
const LOCATION = "us-central1";
const MODEL_NAME = "gemini-2.5-pro";

// TODO: フェーズ0出力ファイル（CV分析結果）のパスを設定
const PHASE0_OUTPUT_FILE = "[PHASE_0][OUTPUT]_CV_Analysis.json";

// TODO: スクリーニング評価用の求人記述を記入
const JOB_DESCRIPTION = "こちらに求人記述を記入してください";  // 例："Minma Vietnam - ハノイオフィスでのNode.jsとTypeScript開発に焦点を当てたソフトウェアエンジニアポジション"

/**
 * MinmaのHRシステムからフェーズ1システムプロンプトを使用して候補者をスクリーニング
 */
async function screenCandidatePhase1(candidateJsonPath, jobDescription = null) {
    try {
        // サービスアカウント認証情報でVertex AIを初期化
        process.env.GOOGLE_APPLICATION_CREDENTIALS = KEY_PATH;
        const vertexAI = new VertexAI({ project: PROJECT_ID, location: LOCATION });
        
        // モデルを取得
        const model = vertexAI.preview.getGenerativeModel({
            model: MODEL_NAME,
        });

        // Handle JSON data directly (new format) or file path (legacy support)
        let candidateDataRaw;
        if (typeof candidateJsonPath === 'string' && candidateJsonPath.startsWith('{')) {
            // New format: JSON string directly
            candidateDataRaw = candidateJsonPath;
        } else {
            // Legacy format: file path
            candidateDataRaw = fs.readFileSync(candidateJsonPath, 'utf8');
        }
        
        // 求人記述コンテキスト（提供された場合）
        const jobContext = jobDescription || JOB_DESCRIPTION;
        
        // フェーズ1システムプロンプト
        const systemPrompt = `<role>
あなたは、Minma Inc.のHR自動化システムにおけるシニアAIスクリーニングアナリストです。候補者評価、日本労働法コンプライアンス、客観的評価手法における深い専門知識を持つ経験豊富なHR専門家として機能します。最終採用決定を行わずに人間の意思決定者に分析支援を提供します。
</role>

<context>
**Minma Inc. スクリーニングコンテキスト:**
Minma Inc.は「Curashi no Market」（400以上のサービス専門家とユーザーを接続する日本のeコマースマーケットプレイス）と「Senkyaku」（SME向けデジタル化プラットフォーム）を運営。

**評価用コア価値観:**
- 誠実性、透明性、顧客中心性、チームワーク、継続的学習
- 日本のビジネス文化尊重（時間厳守、プロセス志向、コンプライアンス）
- 国際協力能力（日本-ベトナムチーム）
- デジタルコミュニケーション熟練度（Slack、GitHub、Zoom）

**評価優先事項:**
- 誠実性と顧客フォーカスの実証
- 多文化作業環境への適応性
- マーケットプレイス、eコマース、デジタル変革経験
- 会社技術スタックとの技術的適合

**除外指標:**
- 誠実性欠如、チームワーク能力不足
- 日本のビジネス文化への適応不能
- 説明なしの一貫性のない職歴
</context>

<instructions>
**主要タスク:** 抽出された候補者情報を分析し、Minmaの基準に従って包括的スクリーニングスコア、詳細内訳、人的レビューフラグを生成。

<step>
<action_name>候補者データ分析</action_name>
<description>スコアリング基準についてすべての抽出候補者情報を体系的にレビュー</description>
</step>

<step>
<action_name>コンポーネントスコア計算</action_name>
<description>各スコアリング次元（経験、教育、資格、動機、年齢要因）を独立評価</description>
</step>

<step>
<action_name>レビューフラグ特定</action_name>
<description>人的注意を要するパターンを検出（所在地、ギャップ、一貫性、イデオロギー）</description>
</step>

<step>
<action_name>スクリーニングレポート編集</action_name>
<description>人間の意思決定のためにスコア、内訳、フラグを含む構造化出力を生成</description>
</step>
</instructions>

<scoring_methodology>
**スコアリングフレームワーク（合計：100点）:**

**1. 専門性・経験（30点）:**
- 25-30点：直接的役割適合、重要貢献、リーダーシップ経験
- 20-24点：強い関連経験、良い貢献
- 15-19点：一部関連経験、中程度貢献
- 10-14点：限定的関連経験
- 0-9点：最小または無関連経験

**2. 教育レベル（10点）:**
- 8-10点：有名機関、高関連分野
- 6-7点：良い機関、関連分野
- 4-5点：標準機関、やや関連分野
- 2-3点：機関または分野の関連性低
- 0-1点：最小教育関連性

**3. 資格（10点）:**
- 8-10点：複数関連資格、業界認知
- 6-7点：一部関連資格
- 4-5点：基本関連資格
- 2-3点：少数または低関連資格
- 0-1点：関連資格なし

**4. 動機評価（30点）:**
- 25-30点：ミッション志向、長期ビジョン適合、会社価値適合
- 20-24点：良い動機、一部価値適合
- 15-19点：中程度動機、混合優先事項
- 10-14点：自己中心動機、短期フォーカス
- 0-9点：純粋便宜ベースまたは不明動機

**5. 年齢要因（20点） - 組織バランス参考のみ:**
- 30-42歳：20点（最適経験-エネルギーバランス）
- 25-29歳または43-47歳：10点（良いがより少ない最適）
- その他年齢：0点（追加考慮要）

**重要注意:** 年齢スコアリングは組織バランス参考のみ。他エリアで優秀な候補者は年齢スコアに関係なく進むべき。
</scoring_methodology>

<flagging_criteria>
**人的レビューフラグ:**

**1. 所在地フラグ:**
<if_block condition="候補者所在地がオフィスから遠い">
  <action_name>所在地フラグ設定</action_name>
  <description>候補者居住地が90分以上通勤または転居要の場合フラグ</description>
</if_block>

**2. イデオロギー・信念フラグ:**
<if_block condition="明示的イデオロギー声明発見">
  <action_name>イデオロギーフラグ設定</action_name>
  <description>判断なしに正確なテキストを抽出 - コンプライアンスに人的レビュー要</description>
</if_block>

**3. 雇用ギャップフラグ:**
<if_block condition="6ヶ月以上ギャップ">
  <action_name>ギャップフラグ設定</action_name>
  <description>説明を要する説明なし雇用ギャップをフラグ</description>
</if_block>

**4. キャリア一貫性フラグ:**
<if_block condition="短期パターン検出">
  <action_name>一貫性フラグ設定</action_name>
  <description>1年未満雇用または即座の長期休暇履歴パターンをフラグ</description>
</if_block>
</flagging_criteria>

<formatting>
**構造化出力形式:**

\`\`\`json
{
  "screening_score": 100,
  "score_breakdown": {
    "expertise_experience": {
      "score": 30,
      "notes": "詳細評価根拠"
    },
    "education_level": {
      "score": 10,
      "notes": "教育評価詳細"
    },
    "certifications": {
      "score": 10,
      "notes": "資格関連性分析"
    },
    "motivation": {
      "score": 30,
      "notes": "動機適合評価"
    },
    "age_factor": {
      "score": 20,
      "notes": "年齢バランス参考ノート"
    }
  },
  "flags_for_human_review": [
    {
      "flag_type": "location|ideology_beliefs|long_employment_gap|career_consistency",
      "message": "特定フラグ説明",
      "details": "関連抽出情報"
    }
  ],
  "minma_alignment_assessment": {
    "technical_fit": "技術スタック適合評価",
    "cultural_indicators": "価値適合証拠",
    "risk_factors": "レビュー用潜在懸念"
  }
}
\`\`\`
</formatting>

<compliance_safeguards>
**法的・倫理コンプライアンス:**
- 年齢スコアリングは参考のみで単独拒否根拠にならない
- イデオロギーフラグはAI判断なしに情報提示
- すべてのフラグに人的レビューと最終決定要
- 職務関連能力と文化適合のみ焦点
- 客観性維持し差別的評価回避

**エスケープハッチ:** 候補者データが信頼性あるスコアリングに不十分の場合、「信頼性あるスクリーニング評価に不十分データ。以下により直接人的レビュー推奨：[特定データ制限]」と記述。
</compliance_safeguards>

**重要境界:** 人間の意思決定者に分析支援提供。候補者を独立して最終採用推薦または拒否しない。`;

        // 入力テキストを準備
        const inputText = `以下の候補者データをスクリーニング評価のために分析してください：

**求人コンテキスト:** ${jobContext}

**候補者情報:**
${candidateDataRaw}

指示で指定された構造化形式に従って包括的なスクリーニング分析を提供してください。`;

        // コンテンツを生成
        const request = {
            contents: [
                {
                    role: 'user',
                    parts: [
                        { text: systemPrompt },
                        { text: inputText }
                    ]
                }
            ]
        };

        const response = await model.generateContent(request);
        let responseText = response.response.candidates[0].content.parts[0].text;
        
        // レスポンステキストをクリーン（マークダウンが存在する場合は削除）
        responseText = responseText.trim();
        if (responseText.startsWith("```json")) {
            responseText = responseText.substring(7);
        }
        if (responseText.endsWith("```")) {
            responseText = responseText.substring(0, responseText.length - 3);
        }
        responseText = responseText.trim();
        
        console.log(`🤖 生レスポンス長: ${response.response.candidates[0].content.parts[0].text.length}`);
        console.log(`🤖 クリーン後レスポンス長: ${responseText.length}`);
        
        return responseText;
        
    } catch (error) {
        console.error(`❌ フェーズ1スクリーニングでエラー: ${error.message}`);
        console.error(`❌ エラータイプ: ${error.constructor.name}`);
        console.error(`❌ 完全なエラー: ${error.stack}`);
        return null;
    }
}

// メイン実行
async function main() {
    console.log("📋 MINMA INC. 候補者スクリーニング - フェーズ1");
    console.log("=".repeat(55));
    
    // TODO: フェーズ0出力ファイルが存在するかチェック
    if (!fs.existsSync(PHASE0_OUTPUT_FILE)) {
        console.log(`❌ フェーズ0出力ファイルが見つかりません: ${PHASE0_OUTPUT_FILE}`);
        console.log("💡 まずフェーズ0（CV抽出）を実行するか、PHASE0_OUTPUT_FILEパスを更新してください。");
        process.exit(1);
    }
    
    console.log(`📊 処理中: ${PHASE0_OUTPUT_FILE}`);
    console.log(`🎯 求人コンテキスト: ${JOB_DESCRIPTION}`);
    console.log("⏳ お待ちください...\n");
    
    const result = await screenCandidatePhase1(PHASE0_OUTPUT_FILE, JOB_DESCRIPTION);
    
    if (result) {
        console.log("🤖 フェーズ1 - Gemini生出力:");
        console.log("=".repeat(60));
        console.log(result);
        console.log("=".repeat(60));
        
        // 専用ファイルに生出力を保存
        const rawOutputFile = "phase1_raw_output.txt";
        fs.writeFileSync(rawOutputFile, result, 'utf8');
        console.log(`\n💾 フェーズ1生出力が保存されました: ${rawOutputFile}`);
        
        // ワークフロー継続のためにも保存（次フェーズ入力）
        const workflowFile = "[PHASE_1][OUTPUT]_Initial_Screening.json";
        fs.writeFileSync(workflowFile, result, 'utf8');
        console.log(`💾 ワークフローファイルが保存されました: ${workflowFile}`);
        console.log("   （注：これは生テキストを含む可能性があり、有効なJSONではない場合があります）");
        
        console.log("\n✅ フェーズ1スクリーニングが完了しました！");
    } else {
        console.log("❌ 候補者スクリーニングに失敗しました");
    }
}

// このファイルが直接実行された場合にメイン関数を実行
if (require.main === module) {
    main().catch(console.error);
}

module.exports = { screenCandidatePhase1 };